### ISSUE TITLE
compartment analysis

### ISSUE THREAD

**Issue by BioEpi** [robinweide/GENOVA#306]
hello,
    I was just wondering whether GENOVA calls cooltools to call compartments.
Best,

**Comment by teunbrand** [comment]
Hi there,

No it uses `base::eigen()` on the observed over expected metric minus 1, per chromosome arm. 

Best,
Teun

**Comment by BioEpi** [comment]
got it.
How can I run saddle function if I provide the compartment scores calculated by cooltools?
I got this error: 
Error in if (attr(explist[[1]], "resolution") != attr(CS_discovery, "resolution")) { : 
  argument is of length zero

**Comment by teunbrand** [comment]
To use custom compartment calls, you need to load the experiment at the same resolution, and ensure that you have a bed-like data.frame with the same structure as the loaded data's index (the `IDX` slot). There also needs to be columns in there with numerical values, where the extra column-names correspond with the names of the loaded data.

``` r
library(GENOVA)

exp <- get_test_data("150k")

custom_compart <- setNames(
  as.data.frame(exp$IDX), 
  c("chrom", "start", "end", "bin")
)
custom_compart[[expnames(exp)]] <- rnorm(nrow(custom_compart)) # dummy values
print(head(custom_compart))
#>   chrom  start    end   bin Hap1_WT_150k # <- experiment name
#> 1 chr21      0 150000 18557    0.2238810
#> 2 chr21 150000 300000 18558   -0.1115281
#> 3 chr21 300000 450000 18559    1.8365315
#> 4 chr21 450000 600000 18560    0.4768462
#> 5 chr21 600000 750000 18561   -0.3371738
#> 6 chr21 750000 900000 18562    0.1041547
```

The return value of `GENOVA::compartment_score()` has some attributes that `saddle()` uses. If you want to run `saddle()` on a custom compartment score, you'd need to ensure that the relevant attributes are present. These are:

* The data's resolution
* Whether the compartment score is signed (synced with active/inactive chromatin)
* The partitioning of the data in arms, represented as run-length encoding. In absence of any arms information, just partitioning by chromosome should work too.

``` r
dummy <- structure(
  list(compart_scores = custom_compart),
  resolution = 150e3,
  signed = TRUE,
  partitioning = rle(custom_compart$chrom)
)

visualise(saddle(exp, dummy))
```

![](https://i.imgur.com/RMkunyd.png)

<sup>Created on 2022-09-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by BioEpi** [comment]
thanks so much. yes, it works.
I have another question:
how to set `dist_thres`? 
Q_LO = 0.025 # ignore 2.5% of genomic bins with the lowest E1 values
Q_HI = 0.975 # ignore 2.5% of genomic bins with the highest E1 values

**Comment by BioEpi** [comment]
great. thanks. 
another question is how to load .mcool file via load_contacts?

**Comment by teunbrand** [comment]
Should be doable with for example `load_contacts(signal_path = "/path/to/file.mcool", resolution = 10000, sample_name = "my_sample")`.

**Comment by BioEpi** [comment]
thanks a lot. yes, it works.
do you know how to remove the grey bar?
<img width="627" alt="image" src="https://user-images.githubusercontent.com/12387462/192828105-80acee4b-c535-451d-a32a-cd45cb8ae19b.png">

**Comment by teunbrand** [comment]
If you do the following (given example above), then the `visualise()` function should interpret everything correctly again:

``` r
sad$saddle <- sad$saddle[!(q1 %in% c(1, 40) | q2 %in% c(1, 40))] # drop highest and lowest bins
sad$saddle <- sad$saddle[, c("q1", "q2") := list(
  match(q1, sort(unique(q1))),
  match(q2, sort(unique(q2)))
)]
visualise(sad)
```

**Comment by teunbrand** [comment]
My guess is that there is some discrepancy between the sign of the compartment score between chromosomes, but I'm not familiar enough with cooltools' compartment scores, nor your code to use it with with GENOVA's `saddle()` function, to have an informed opinion.

**Comment by BioEpi** [comment]
there is some discrepancy between the sign of the compartment score between chromosomes--- would you like to tell me how GENOVA assign the sign of the compartment score between chromosomes? negative value is B, while positive is A, right?

**Comment by teunbrand** [comment]
Yes, but only if the compartment score is signed. See also the note in `?compartment_score()`.

**Comment by BioEpi** [comment]
here is the command: saddle(list(a1,a2),CS_discovery=b,bins=40). I am still confused about the huge difference between GENOVA and Cooltools.

**Comment by teunbrand** [comment]
Did you give the compartment scores calculated by cooltools to the saddle function, or compartment scores calculated by GENOVA to the saddle function? If it is the first, I don't really have anything useful to add because I'm not familiar with how cooltools calculates compartment scores. If it is the second, I'm a bit stumped because I'd agree with you that it isn't what typical saddle plots look like.
